framework/model/source/aggregate/metric_source.go

```diff
@@ -0,0 +1,191 @@
+	return nil
+}
+
+func Remove(m source.MetricSourceForAggregate) error {
+	// stop watchers if existing gvk watcher channel
+	if len(metricSource.gVKStopChanMap) > 0 {
+		metricSource.StopWatchers()
+	}
+	// update SourcesMap
+	if _, ok := metricSource.sourcesMap[m.Name()]; ok {
+		delete(metricSource.sourcesMap, m.Name())
+	} else {
+		return errors.New(fmt.Sprintf("metric source %s not exists", m.Name()))
+	}
+	// check if this is the last metric source
+	if len(metricSource.sourcesMap) == 0 {
+		log.Infof("no metric source now, will return")
+		return nil
+	}
+	// update GVKSourceMap and GVKStopChanMap
+	metricSource.MapRemove(m)
+	metricSource.StartWatchers()
+	return nil
+}
+
+func (ms *MetricSource) MapAdd(m source.MetricSourceForAggregate) {
+	for _, gvk := range m.GVKs() {
+		if _, ok := ms.gVKSourceMap[gvk]; !ok {
+			log.Debugf("add new key %s to aggregate metric source gVKSourceMap", gvk.String())
+			// update GVKSourceMap
+			ms.gVKSourceMap[gvk] = []source.MetricSourceForAggregate{m}
+			// update GVKStopChanMap
+			ms.gVKStopChanMap[gvk] = make(chan bool)
+		} else {
+			log.Debugf("update key %s in aggregate metric source gVKSourceMap", gvk.String())
+			// only update gVKSourceMap, GVKStopChanMap[gvk] already exists
+			ms.gVKSourceMap[gvk] = append(ms.gVKSourceMap[gvk], m)
+		}
+	}
+}
+
+func (ms *MetricSource) MapRemove(m source.MetricSourceForAggregate) {
+	for _, gvk := range m.GVKs() {
+		if s, ok := ms.gVKSourceMap[gvk]; !ok {
+			log.Warningf("aggregate metric source gVKSourceMap does not contain key %s", gvk.String())
+		} else {
+			for i, v := range s {
+				if v.Name() == m.Name() {
+					if len(s) == 1 {
+						// the last user of this gvk, delete key
+						delete(ms.gVKStopChanMap, gvk)
+						delete(ms.gVKSourceMap, gvk)
+					} else {
+						// only update GVKSourceMap, GVKStopChanMap[gvk] still has other users
+						s = append(s[:i], s[i+1:]...)
+					}
+					log.Debugf("successfuly remove gvk %s of metric source %s", gvk.String(), m.Name())
+					break
+				}
+			}
+			log.Errorf("failed to remove gvk %s of metric source %s", gvk.String(), m.Name())
+		}
+	}
+}
+
+func (ms *MetricSource) StopWatchers() {
+	for _, ch := range ms.gVKStopChanMap {
+		go func() {
+			ch <- true
+		}()
+	}
+	log.Infof("Aggregate MetricSource sends stop signals to all watchers")
+}
+
+func (ms *MetricSource) StartWatchers() {
+	// start MetricWatcher
+	var err error
+	ms.watcherEventChan, err = watcher.Register(ms)
+	if err != nil {
+		log.Errorf("MetricSource registers to MetricWatcher error: %v", err)
+		return
+	}
+	ms.ConsumeEvent()
+}
+
+// 消费WatcherEvent: 提炼出gvk+nn信息，找出关联的modules
+func (ms *MetricSource) ConsumeEvent() {
+	// check sources number
+	go func() {
+		for {
+			if len(ms.sourcesMap) == 0 {
+				log.Infof("mo metric source in sourcesMap, stop consuming watcherevent...")
+				ms.stopChan <- true
+				return
+			}
+		}
+	}()
+
+	// create fake watcher event from ticker
+	ticker := time.NewTicker(30 * time.Second)
```

Comment:
timer event should be registered/triggered according to the config/requirement from users